### ISSUE TITLE
full testing required before code files modified?

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/NCEPLIBS-bufr#480]
We are now pretty close to full testing.

Do we agree that it's reasonable to start imposing a new process rule concerning test code?

Before any file is modified, it must be brought up to full test coverage.

By "full" we mean as close to 100% as is reasonable. As we see, many codes can reach 100%. Sometimes statements really can't be tested, but I don't see any examples in our code where 95%+ is not readily achievable.

This may be relaxed for emergencies, but should cover all new development.

What do you think @jbathegit ?

**Comment by jbathegit** [comment]
What I think is that #79 is important too.  As I've noted previously, the next release is already going to be a major release because of the consolidation down to one integer\*4 build, which in turn requires any integer\*8 application codes to change to now use the new setim8b.  And we now have a request (see #462) for a new release sooner rather than later, so if we don't get #79 done as part of this new release, then the next release after that will also need to be yet another major release, since #79 will require any C application codes to change to use the new interface to the Fortran part of the library.  So it makes sense to get #79 done within the same release and just have one major release.

I get that testing is important, and as you know I've contributed a lot towards that effort during the past several months.  But #79 also needs to get done as part of this next major release.  I've already put it on the backburner several times to work on other efforts that were deemed more important, such as getting everything fully Doxygen documented, and then getting rid of all of the compiler warning and address sanitizer issues, and then improving the overall testing.  So at this point I don't see how I can put this off any longer, and it's not like I'm adding any untested code to the base.  Most of the work for #79 involves wrappers to change the interface between C and Fortran parts of the library, and if there is any added code I've been making sure that testing is included.  In short, I don't see why I can't work on #79 while we continue to bring the overall testing up to 95%+ at the same time.

**Comment by edwardhartnett** [comment]
As an alternative suggestion, how about:

1. For work on #79 take a look at the files you are modifying and put new issues up for any untested code. Alex and I will jump in and help, and if you and we all spend just one day writing test code, I'd be surprised if we couldn't make significant progress. 
2. After #79 and the 12.0 release, we agree that no further code modifications will take place, except on fully tested code (emergency bug fixes can be excepted from this rule).

Playing catch-up on testing is some effort, as we see. Let's ensure that the processes we have in place properly require that test code be written by the original programmer, and that no code can be committed without complete testing. This is new to NOAA, but long-established custom in almost all professional programming organizations.

Culture changes are hard, but we cannot proceed with 1980s techniques in the 21st Century. If we want NCEPLIBS-bufr to match the quality and productivity of other major I/O libraries, we must make testing central to the programming and debugging effort. 

I am struggling to understand how you can even be modifying code that is not fully tested. I would be terrified to do so, and would regard writing the tests first as the fastest way to accomplish anything. How do you even know your changes are working?

**Comment by jbathegit** [comment]
I hear what you're saying Ed, but again please understand that I'm never adding any untested code!

Almost all of this C-Fortran interface is already being used inside of the library (e.g. cadn30, nemtab, igetntbi), so if I update one of those wrappers, I'm correspondingly updating the interface within the library to now use that new wrapper.  So that's how I know the updated wrapper is continuing to work properly.  And for the handful of wrappers that weren't already being called inside of the library (e.g. isetprm, maxout) I've been adding appropriate tests to the test/test*c codes at the same time.  So that's how I know those wrappers are also continuing to work properly.

Bottom line, I don't see how I'm adding any untested code or otherwise possibly breaking something that's already working, so I really don't see an issue here(?)

**Comment by jbathegit** [comment]
That said, I will also agree with your point 2 above, that beyond #79 and the upcoming 12.0.0 release, we don't add any new non-emergency changes to the library until we're all satisfied with the level of testing.  I still hope we can get as much of that knocked out between now and then anyway.

**Comment by edwardhartnett** [comment]
One line in seven is completely untouched in testing:

![image](https://user-images.githubusercontent.com/38856240/236254597-c5da89b7-1600-43c1-824b-957bb07eb891.png)

Which code files do you intend to edit next? Can you identify them so we can ensure they are tested?

**Comment by jbathegit** [comment]
Most of my non-documentation changes (where I'm actually updating wrappers) are in bufr_c2f_interface.F90, bufrlib.h.in, and bufr_interface.h.  So all good there.  And then I'm correspondingly updating the lines in the library where those wrappers are being called (e.g. restd.c, stseq.c) to now use those new wrappers instead of the old ones.  Like I said, I'm not changing any functionality, just using/testing the updated wrappers.

So how about this - please feel free to go through any of my recent and upcoming PRs with a fine-tooth comb, and if you spot anything I'm changing that's not being tested or which you feel could in any way be compromising the integrity of the existing code, then by all means feel free to call me out on it.  Fair enough?

**Comment by edwardhartnett** [comment]
It takes more programmer time to maintain untested code than it does to write tests and maintain tested code. This is a well-proven engineering fact, supported by endless research. 

I don't think we want to change any code that is not fully tested.

This does not place much of an extra burden on programmers, even in the short term. We have demonstrated that we can quickly write tests, targeted to code that is untested, using code coverage reports. For example, in your current PR, I noticed an untested line of code and was able to add a test for it in less than 10 minutes (#489).

We are trying to raise the bar of all NCEPLIBS programming processes to industry standard. That requires changing how we do things. Probably the hardest cultural change is away from NOAA toleration of untested code. 

If code is untested, how do we know it works? 

I think it's quite reasonable and very normal for programmers to take a pass at test code before making any changes to production code. Simply look at the code report for the codes you want to modify, and see if you can add testing. It's really very easy - the easiest kind of programming there is. If you need help, ask for it, and Alex and I can chip in.

**Comment by edwardhartnett** [comment]
Shall we close this question?

I think it's great the way that testing has proceeded on NCEPLIBS-bufr, and is still being improved long after our initial test campaign ended. This is the sign of a project that has successfully made the transition to test-driven development!

I think it is now reasonable that if changes were to be needed in any of the few remaining untested areas of code, that it would be a good idea to first fill in those tests, which was the origin point of this question. @jbathegit do you agree?

I would venture to say that NCEPLIBS-bufr is the best documented, best tested code at the EMC right now, probably in all of NOAA.

**Comment by jbathegit** [comment]
Agreed, and thanks!
